nemo/src/program_analysis/type_inference/position_graph.rs

```diff
@@ -0,0 +1,319 @@
+    BodyToHeadAggregateStaticOutputType,
+    /// Describes data flow from an aggregate input variable to and aggregate output variable, where the aggregate output variable has the same type as the input variable (see [`AggregateOperation::static_output_type`]).
+    BodyToHeadAggregateNonStaticOutputType,
+}
+
+pub(super) type PositionGraph = LabeledGraph<PredicatePosition, PositionGraphEdge, Directed>;
+
+impl PositionGraph {
+    pub(super) fn from_rules(rules: &Vec<ChaseRule>) -> PositionGraph {
+        let mut graph = PositionGraph::default();
+
+        for rule in rules {
+            let mut variables_to_head_positions =
+                HashMap::<Variable, Vec<PredicatePosition>>::new();
+
+            for atom in rule.head() {
+                for (term_position, term) in atom.terms().iter().enumerate() {
+                    if let PrimitiveTerm::Variable(variable) = term {
+                        let predicate_position =
+                            PredicatePosition::new(atom.predicate(), term_position);
+
+                        variables_to_head_positions
+                            .entry(variable.clone())
+                            .and_modify(|e| e.push(predicate_position.clone()))
+                            .or_insert(vec![predicate_position]);
+                    }
+                }
+            }
+
+            let mut aggregate_input_to_output_variables =
+                HashMap::<Variable, Vec<(Variable, PositionGraphEdge)>>::new();
+            for aggregate in rule.aggregates() {
+                for input_variable_identifier in &aggregate.variables {
+                    let edge_label = if aggregate.aggregate_operation.static_output_type().is_some()
+                    {
+                        PositionGraphEdge::BodyToHeadAggregateStaticOutputType
+                    } else {
+                        PositionGraphEdge::BodyToHeadAggregateNonStaticOutputType
+                    };
+
+                    aggregate_input_to_output_variables
+                        .entry(input_variable_identifier.clone())
+                        .or_default()
+                        .push((aggregate.output_variable.clone(), edge_label));
+                }
+            }
+
+            let mut variables_to_last_node = HashMap::<Variable, PredicatePosition>::new();
+
+            for atom in rule.all_body() {
+                for (term_position, variable) in atom.terms().iter().enumerate() {
+                    let predicate_position =
+                        PredicatePosition::new(atom.predicate(), term_position);
+
+                    // Add head position edges
+                    {
+                        // NOTE: we connect each body position to each head position of the same variable
+                        if let Some(head_positions) = variables_to_head_positions.get(variable) {
+                            for pos in head_positions {
```

Comment:
Just an idea: maybe this can be written as a single loop using `flatten()`, to remove one layer of indentation like so:
```rust
let head_positions = variables_to_head_positions.get(variable);
for position in head_positions.into_iter().flatten() {
  ..
}
```

Reply:
Ah yes, I never really considered the possibility of turning an `Option` into an iterator but this makes a lot of sense :)